[RFC] osd: tier flush potentially leading to inconsistent

Aran85 opened the pull request:
When flush clone object, we will generate delete op, this uesd to let
head in base pool make clone correctly. but the delete op is send from
tier pg to base pg, so if tier pg partition happen, it will still send
to base pg. so we need extra info to avoid the case.

Detail:

Let's say tier pg has a partited one called pg[1,2]p1, p1 is disconnected
with others and monitor, so it will not receive new osdmap, and keep
active state. the auth pg is already become pg[2,3]p2 ,and start to flush.

When osd.1 reconnect and before advance to new osdmap, the agent_work
may start to flush, and send delete and copy_from to base pool, delete
will be handle and remove head, but copy_from will fail, because the flush
already occured.. and the head in base pool will be deleted wrongly and
can not recovery.

Change:
1. I introduce a new member called flushed_thru into pg_info_t. 
2. delete and copy_from generated when do flush will take the flag of CEPH_OSD_FLAG_FLUSH to
    let it be distinguishable.
3.  when these ops processing, we update the pg_info_t flushed_thru.
4.  If we receive the ops with older epoch than current flushed_thru, reply error.

Fixes: https://tracker.ceph.com/issues/38787

Signed-off-by: Zengran Zhang <zhangzengran@sangfor.com.cn>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

gregsfortytwo commented:
I think I see what you're saying, but can you describe it in a bit more detail? What's the state in the cache and base pools, what is the exact breakage, and why does this change fix it?

Aran85 commented:
@gregsfortytwo sorry for respond late, I will try to describe it in a bit more detail.

The issue could easily verify. make a writeback mode tier pool on cephfs_data, create a file, then cache-flush-evict-all the tier pool, checkout the base osd log, we will find 2 oprations, 
1) delete with snapcontext seq 0
2) copy from with snapcontext seq 1
Because there is no object in the base pool, so it is an error when delete non-exist object, record a pglog of error. here is the issue, there should not been a delete op..

Then if we rewrite the file, the object will promote to tier pool again, but meanwhile remain in base pool. if we cache-flush-evict-all tier pool again,then there will be 2 ops as above, but this time deletion works, but unnecessarily and not atomic..

It seems the head object should not be delete by single delete on the code which I changed, copy-from will remove the object in pgtransaction.(see PrimaryLogPG::finish_copyfrom())

The breakage I think maybe cause by cephfs start  snapcontext from 1, but osd use non-zero to decide there must be clones. the code which I changed I think could avoid extra deletion when there is no clones.

Does that make sense?

Aran85 commented:
retest this please

Aran85 commented:
@liewegas @athanatos I misunderstanded the delete op, the issue I had amended on above. the cephfs start snapset seq from 1 will not cause data inconsistent. (but I still think this will cause no snapshot cephfs environment will try to delete head first, and seems like it can not avoid).

I want to fix the potential inconsistent possibility when tier flushing object in this PR. To descript the issue clearly is really a challenge for me, if I have not, I will try do better.

athanatos commented:
I don't understand, the delete ought to be idempotent and the flush won't be considered complete until both happen.  The rados operation itself would retry within the objecter if there's an interval change on the base tier primary.  Can you explain a bit more about what you actually observed?

Aran85 commented:
@athanatos I observe the object had finished flush, then an prev primary osd(I call it partited pg) send delete+copy_from again, the delete op remove the object but copy_from failed. here the delelte and copy_from are not atomic, and delete will not assert the object version which going to remove.

Did you had read the comment at the beginning? It's a little hard for me to describe the circumstance, but seems that you did not read my comment... :<

athanatos commented:
Hmm.  I think I see what's going on.  First, let me type the sequence back out so you can tell me if I've got it right.

pool 2 is cache for pool 1 (same number of pgs)
1) pg2.1 e7 osd 1 primary sends delete + copy_from on foo or pg1.1 osd 11
2) osd 1 gets marked down (but isn't really partitioned)
3) osd 2 becomes primary for pg 2.1 in e8 and sends delete + copy_from
4) pg 1.1 osd 11 e8 processes delete + copy_from from pg2.1 osd 2 e8
5) pg2.1 osd 2 deletes foo as the flush has completed
6) pg 1.1 osd 11 e8 processes delete + copy_from from pg2.1 osd 1 e7

At step 5, foo is deleted but the copy_from fails because the object has been removed from the cache pool.

However, your fix has a problem.  Suppose the cache pool has twice (or 4x, or 8x, ...) the pgs as the base pool.   The objects in the base pool would be flushed by more than one single pg and thus more than one set of primaries.  A single pg-wide value would incorrectly reject flushes in some cases when the different cache primaries are at slightly different epochs.  I think the correct approach would be for the flushing operation to maintain an atomic epoch counter in an xattr on the base tier object.  The delete would be sent guarded with a cmpxattr and the copy_from would be sent with a subsequent setxattr.

Does that make sense?

As I mentioned on your other PR, patches like this are much much much easier to merge if you can find a way to create a standalone test in qa/standalone/osd to ensure that your fix works and that we won't rebreak it in the future.

athanatos commented:
Another possibility is that we could check whenever we get a rados operation from an osd whether the osd's down_at is after the epoch of the MOSDOp.  That actually sounds way easier.

Aran85 commented:
> A single pg-wide value would incorrectly reject flushes in some cases when the different cache primaries are at slightly different epochs

this means we may reject some legal flush with older epoch, but they will retry, right?

> Another possibility is that we could check whenever we get a rados operation from an osd whether the osd's down_at is after the epoch of the MOSDOp. That actually sounds way easier.

but sounds that will affect more generally..

athanatos commented:
> > A single pg-wide value would incorrectly reject flushes in some cases when the different cache primaries are at slightly different epochs
> 
> this means we may reject some legal flush with older epoch, but they will retry, right?

Only if the flush logic were modified to make it so.  I really think this approach is both brittle and hard to implement.

> 
> > Another possibility is that we could check whenever we get a rados operation from an osd whether the osd's down_at is after the epoch of the MOSDOp. That actually sounds way easier.
> 
> but sounds that will affect more generally..

I can't think of any case where we actually want to process an op from an osd we know to be dead.

Aran85 commented:
> > > A single pg-wide value would incorrectly reject flushes in some cases when the different cache primaries are at slightly different epochs
> > 
> > 
> > this means we may reject some legal flush with older epoch, but they will retry, right?
> 
> Only if the flush logic were modified to make it so. I really think this approach is both brittle and hard to implement.
> 
seem that tier agent will keep scanning..

> > > Another possibility is that we could check whenever we get a rados operation from an osd whether the osd's down_at is after the epoch of the MOSDOp. That actually sounds way easier.
> > 
> > 
> > but sounds that will affect more generally..
> 
> I can't think of any case where we actually want to process an op from an osd we know to be dead.

I will see how to walk on this way :D
